lib/srv/desktop/ldap.go

```diff
@@ -33,17 +34,28 @@ type ldapClient struct {
 // newLDAPClient connects to an LDAP server, authenticates and returns the
 // client connection. Caller must close the client after using it.
 func newLDAPClient(cfg LDAPConfig) (*ldapClient, error) {
-	con, err := ldap.Dial("tcp", cfg.Addr)
-	if err != nil {
-		return nil, trace.Wrap(err)
+	tlsConfig := &tls.Config{
+		InsecureSkipVerify: cfg.InsecureSkipVerifyCA,
 	}
-	// TODO(awly): should we get a CA cert for the LDAP cert validation? Active
-	// Directory Certificate Services (their managed CA thingy) seems to be
-	// issuing those.
-	if err := con.StartTLS(&tls.Config{InsecureSkipVerify: true}); err != nil {
-		con.Close()
+
+	if !cfg.InsecureSkipVerifyCA {
```

Comment:
I think this if should be `if cfg.CA != nil`.

The LDAP server cert may already be in the system's trust store, in which case there would be no need to set the CA cert in the configuration.

Reply:
As written I'm ignoring the system certs and only using the one provided (https://github.com/gravitational/teleport/blob/isaiah/windows-ldaps-port/lib/srv/desktop/ldap.go#L42-L44).

I went with this for now, since I found I needed to [provide a `tlsConfig.ServerName`](https://github.com/gravitational/teleport/blob/isaiah/windows-ldaps-port/lib/srv/desktop/ldap.go#L50-L52) in order for the verification to work. However if I'm using the system cert pool, I couldn't find a good way to provide that.

I'm not 100% clear why this works like it does -- maybe I wouldn't need to provide `ServerName` for a system cert (?). I can experiment with that.

Reply:
In case it helps -- the error when I don't provide `tlsConfig.ServerName` is
```
Original Error: *ldap.Error LDAP Result Code 200 &#34;Network Error&#34;: x509: cannot validate certificate for 172.18.0.4 because it doesn&#39;t contain any IP SANs
```

Reply:
Ah yes, it does seem to work if I add the cert to my system's trusted pool. I don't know why, but I'll update to allow for this type of configuration.

Reply:
It doesn't work if I add the cert to the system cert pool. I think this is happening because I am calling to `172.18.0.4`, but the cert is issued for `winserver2016.zacazure.grvpoc.com`. I'm wondering now if setting `tlsConfig.ServerName = cfg.CA.Subject.CommonName` is a mistake, and if we should instead rely on the certificate to have the appropriate SAN.

Reply:
Right, the certificate should have the appropriate SAN(s). In most real world use cases, the `addr` will likely be a hostname, not the internal IP of a single DC.